packages/services/api/src/modules/audit-logs/providers/audit-log-recorder.ts

```diff
@@ -0,0 +1,77 @@
+import { randomUUID } from 'node:crypto';
+import { Injectable, Scope } from 'graphql-modules';
+import { captureException } from '@sentry/node';
+import { type User } from '../../../shared/entities';
+import { sql as c_sql, ClickHouse } from '../../operations/providers/clickhouse-client';
+import { Logger } from '../../shared/providers/logger';
+import { auditLogSchema, type AuditLogSchemaEvent } from './audit-logs-types';
+
+type AuditLogRecordEvent = AuditLogSchemaEvent & {
+  metadata: {
+    userId: string;
+    userEmail: string;
+    organizationId: string;
+    user: User;
+  };
+};
+
+@Injectable({
+  scope: Scope.Operation,
+  global: true,
+})
+/**
+ * Responsible for recording audit log events and storing them in ClickHouse
+ */
+export class AuditLogRecorder {
+  private logger: Logger;
+
+  constructor(
+    logger: Logger,
+    private clickHouse: ClickHouse,
+  ) {
+    this.logger = logger.child({ source: 'AuditLogRecorder' });
+  }
+
+  async record(data: AuditLogRecordEvent): Promise<void> {
+    try {
+      const { eventType } = data;
+      const { organizationId, userEmail, userId } = data.metadata;
```

Comment:
This is a bit messy to me. We have metadata, that does not land in the `metadata` column.

Reply:
@TuvalSimha if some props are shared across all events, you could create a space for them, a `session` or `context` property with `{ userId, userEmail, organizationId }` in it, all "unique" properties can be in its own space.

Then when you write data to `metadata` column, it's safe as it was previously parsed with Zod (no data leaks, no additional properties).

When you write to `organization_id` or `user_id` columns, you will use `context.user_id` or something.

By passing only necessary data to those events, you have less complexity, less chances for data leak or errors.

Reply:
Props could be at top level as well, as long as you parse the data correctly.

Reply:
I changed it, and parsed it with Zod schema for both: metadata and the context:

```
parsedMetadata {
  eventType: 'TARGET_CREATED',
  metadata: {
    projectId: '0cc37d44-ed94-4467-a3f6-f588c3e3742f',
    targetId: '6d745f0f-58f2-439e-81e1-37ce26ff4e7e',
    targetSlug: 'development'
  }
}
parsedContext {
  fullName: 'Tuval Simha',
  email: 'tuval.simha@gmail.com',
  displayName: 'Tuval Simha',
  provider: 'USERNAME_PASSWORD'
}
```